### PR TITLE
Refactor battle interface to add a dedicated surface

### LINKED ISSUE #191: Battle animation goes out of battle window
Run the game in resolution more than 640 x 480 pixels and run any battle. Move unit close to upper side of battle window and notice that images stay on world map until the end of the battle which looks very weird:

Screenshot 2019-10-22 at 5 47 23 PM

### LINKED ISSUE #778: Pieces of journal scrollbar can be seen on a battlefield
While moving of enemy unit try to click the right bottom corner of battlefield, You can see there a small piece of scrollbar for a moment.
Battlefield bug

### LINKED ISSUE #804: Battlefield objects are drawn below the battle in FullHD resolution
Start any battle in 1920x1080 resolution and notice this:
image

### PR DESCRIPTION
**Pull request by idshibanov**
Fixes #191 . Fixes #804 . Fixed #778 

Main idea here is to draw on a separate surface instead of directly on `Display`. Also a lot of refactoring to split battle interface draw into two parts and get rid of unnecessary references to cursor and display.

### CODE REVIEW

File: src/engine/rect.h [hunk 1]
```diff
@@ -79,14 +79,19 @@ struct Rect : Point, Size
...
+    bool operator&( const Point & ) const;
     // rect intersects rect
-    bool operator&(const Rect &)const;
+    bool operator&( const Rect & ) const;
 
     // calculate intersection rectangle
     Rect operator^( const Rect & other ) const;
 
+    // explicit conversion
+    const Point & getPosition() const;
```
Comment (idshibanov) [inline src/engine/rect.h:93]: Making an explicit call here since `Rect + Rect` works, but is very ambiguous.

**Comment by idshibanov** [comment]
@ihhub Okay, I think I'm done here. A lot of changes to review, but most of them changing the same thing. I hope with new variables names it's not too hard to read.

Tested full HD resolution: looks good to me, however game is very slow on Windows. I did not profile the performance, might take a look later separately.

On a side note, `clang-format-diff` script is acting weird: look at my last commit, that's the format it wanted. Why?

**Comment by ihhub** [comment]
Hi @idshibanov , I am just curious whether we could have a simpler way to do it. What if we trick Display class? For example, we can set a temporary Surface within Display class and use it inside `void Surface::Blit( const Rect & srt, const Point & dpt, Surface & dst ) const` method to draw on it instead of Display and then inside Flip method we Blit the surface on Display? Think about this idea as a display within display :)

**Comment by idshibanov** [comment]
> Hi @idshibanov , I am just curious whether we could have a simpler way to do it. What if we trick Display class? For example, we can set a temporary Surface within Display class and use it inside `void Surface::Blit( const Rect & srt, const Point & dpt, Surface & dst ) const` method to draw on it instead of Display and then inside Flip method we Blit the surface on Display? Think about this idea as a display within display :)

What I wanted to achieve here is to update all drawing positions to be relative, not global to the window. It will be easier to scale up the battle window later. Also I wanted to get rid of all `Display::Get()` and `display.Flip()` calls in every drawing function.

How do we control this inner surface with your approach; determine which one to use? There's still a lot of code around drawing directly to the screen, especially in interface.

**Comment by ihhub** [comment]
> > Hi @idshibanov , I am just curious whether we could have a simpler way to do it. What if we trick Display class? For example, we can set a temporary Surface within Display class and use it inside `void Surface::Blit( const Rect & srt, const Point & dpt, Surface & dst ) const` method to draw on it instead of Display and then inside Flip method we Blit the surface on Display? Think about this idea as a display within display :)
> 
> What I wanted to achieve here is to update all drawing positions to be relative, not global to the window. It will be easier to scale up the battle window later. Also I wanted to get rid of all `Display::Get()` and `display.Flip()` calls in every drawing function.
> 
> How do we control this inner surface with your approach; determine which one to use? There's still a lot of code around drawing directly to the screen, especially in interface.

We could create some kind of singleton which could hold an image. At the image of calling Blit function we "secretly" swap display with this image if display is bigger than the image and then use the image upon Flip method call. However, we have to 'register' this object for the time of this battle window duration.

**Comment by ihhub** [comment]
Another way to do the same is to have a separate function to drawing where we exactly use the image or Display based on game window size. In this case we could easily control what we want. This solution is good but then we have another window for castle which needs the same behavior... I'm just thinking how to make it simpler for the future changes. What do you think?

**Comment by idshibanov** [comment]
> We could create some kind of singleton which could hold an image. At the image of calling Blit function we "secretly" swap display with this image if display is bigger than the image and then use the image upon Flip method call. However, we have to 'register' this object for the time of this battle window duration.

Okay, I understand; it will work. However I'm not sure what are the benefits of doing it in display? Battle window is a sort of "game within the game" that's why it makes sense to me to have it's own surface. Do you want to reuse this approach for other dialogs?

**Comment by idshibanov** [comment]
One problem that stands out with doing it in the `Display`: if we apply same approach to multiple windows, they will conflict with each other. Think about battle window registering own smaller surface and then spellbook widget making another register call as well.

We can of course limit which dialogs use this approach but it's a clean universal solution.

**Comment by ihhub** [comment]
Hi @idshibanov , I pulled your changes and ran with SDL 1. I have a pretty interesting image:
image
:)

**Comment by idshibanov** [comment]
> 
> 
> Hi @idshibanov , I pulled your changes and ran with SDL 1. I have a pretty interesting image :)

Oh, fun times with SDL 1 alpha. And all of that is fixed with simple flip of `true` to `false`

**Comment by ihhub** [comment]
Hi @idshibanov , I noticed that Cold Ray spell has issues as well:
image

**Comment by ihhub** [comment]
Hi @idshibanov , I think these issues are related to your changes:
Run the application in default (640x480) resolution and notice this:
image

image

**Comment by idshibanov** [comment]
@ihhub On which OS it happened? Tried to reproduce this morning but couldn't.. both SDL1/SDL2 in different resolutions, win/lose/retreat screens. That is very odd.

**Comment by ihhub** [comment]
> @ihhub On which OS it happened? Tried to reproduce this morning but couldn't.. both SDL1/SDL2 in different resolutions, win/lose/retreat screens. That is very odd.

Hi @idshibanov , I tested on Windows 7. I will try to test again.

**Comment by ihhub** [comment]
Hi @idshibanov , I tested again. The problem with cold ray still persist while coloring is gone. I'm not sure what happened previous time.

**Comment by ihhub** [comment]
I also noticed that with lightning animation this solution crashes in` DrawLine` function. I've created a separate pull request to fix it (#905 ).

**Comment by idshibanov** [comment]
> 
> 
> Hi @idshibanov , I tested again. The problem with cold ray still persist while coloring is gone. I'm not sure what happened previous time.

@ihhub  Got it! That was because it was calling display flip directly in the `RedrawRaySpell` missing interface update. Should be handled now. Also corrected some code in lighting bolt call.

**Comment by ihhub** [comment]
Hi @idshibanov , could you please help me to fix this merge conflict?

**Comment by ihhub** [comment]
Hi @idshibanov , my biggest concern of this solution is that there is a chance that someone else or one of us will forget about usage of extra surface and use **Display** directly. We should have some sort of lock to prevent using Display for drawings. I'm thinking how to do it in the best way.

**Comment by idshibanov** [comment]
> 
> 
> Hi @idshibanov , my biggest concern of this solution is that there is a chance that someone else or one of us will forget about the usage of extra surface and use **Display** directly. We should have some sort of lock to prevent using Display for drawings. I'm thinking how to do it in the best way.

I understand this, however blitting directly on display like it is now has similar drawbacks too - you can forget to apply window offset and issue similar to #804 will happen again. Such functions as `RedrawCoverStatic` and `RedrawCastle` in this file already drawn on passed subsurface reference; it was a mix of two approaches. If you can come up with the way to make drawing target more explicit - I'm all for it.

Merge conflict should be fixed, I tested a bit with Armageddon and Chain Lightning spells.

**Comment by ihhub** [comment]
@idshibanov , thanks a lot for fixing these issues! Great work as usual!

**Pull request merged**